Introduces contextual init-population if arc is not empty.

piotrswigon opened the pull request:
Summary:
* Planificator controls whether planning is contextual (i.e. targeted to arc content) or broad (i.e. includes all context recipes).
* InitPopulation can fetches either all recipes from the RecipeIndex or only those matching active recipe slots or handles.

mmandlis commented on runtime/strategies/arc-content-matches.js:
```diff
@@ -0,0 +1,50 @@
+      assert(handle.id, 'Handle in the active recipe should have an id');
+      let store = null;
+      for (let otherHandle of this._arc.recipeIndex.findHandleMatch(handle, ['use', '?'])) {
+        if (!store) store = this._arc.findStoreById(handle.id);
```
nit: can you just 
```
let store = this._arc.findStoreById(handle.id);
assert(store);
```
outside the for loop?

piotrswigon commented:
Conceptually I think we should always do (2), though today that might yield different results from (1) due to some pre-processing done by the index. We should fix it and it's on my plate.

I agree with you that it would be nicer to have all the use case served by a single strategy - will look into it and get back to you.

Thank you for feedback :)

piotrswigon commented:
@mmandlis I rewrote this patch as suggested.

(1) -> (2) was done in #1569 and this patch now performs (3) inside InitPopulation, as you recommended.

I realized that I need to also find recipes that can consume existing slots, so I'm waiting for your #1560 to add this, otherwise webdriver tests fail.

One change I realized I needed to make was to make Planificator responsible for choosing between (2) and (3). This is because planificator is responsible for managing relationship between search and planning. As long as there is no search term and showAll is false, (3) is fine. However if we have showAll or a search term, we today need (2) to get all results. This requires the planificator to understand whether the results it holds are contextual (i.e. (3)) or not (i.e. (2)), so I modified the planificator, but maybe you can advise how to improve it in the meantime.

I will need to add bunch of tests to this patch, as well, will come soonish.

mmandlis commented:
it seems to me that you could simplify the planificator changes a tiny bit -
there are 3 events that can trigger replanning today: search string change, plan instantiation and data changes.
* when search is set to anything non-empty (* or a string) we don't want contextual search.
* we do want contextual search in the other cases - when a plan was instantiated or data was changed.
so we can just pass the `contextual` to _requestPlanning from the calling method, and no need for the `_shouldRequestContextualPlanning` method or to store the `contextual` in the results set. 
I think. Unless i'm missing something.

piotrswigon commented:
I think we need to remember whether we did non-contextual planning.

E.g. if the search term changed to `*`, then if there are only contextual results available we should re-plan with `contextual: false`, but if we already did non-contextual planning earlier, we only need to modify the search filter and no re-planning is needed.

Also, if data change or plan instantiation happened and there is currently a search term present, we need to replan non-contextually again with new data.

WDYT?

piotrswigon commented:
FYI I'll be adding various tests to this now.

piotrswigon commented:
I think this is now ready for review :) @mmandlis 

This will unlock enabling CoalesceRecipes again :)

mmandlis commented on runtime/test/strategies/init-population-tests.js:
```diff
@@ -58,8 +59,94 @@ describe('InitPopulation', async () => {
+      // Binds to slot tortilla
+      particle FillsTorilla
+        consume tortilla
+      recipe FillsTorilla
```
typo - FillsTorilla -> FillsTortilla

piotrswigon replied:
Fix, thanks :)

mmandlis commented on runtime/test/recipe-index-test.js:
```diff
@@ -132,4 +154,184 @@ describe('Recipe index', function() {
+    let index = await createIndex(`
+      schema Thing
+
+      particle Conumer
```
typo - Conumer -> Consumer

piotrswigon replied:
Fix, thanks :)
